src/core/move_handling.h

```diff
@@ -103,16 +110,22 @@ constexpr void performPromotionMove(BitBoard& newBoard, movegen::Move move)
     // first clear to be promoted pawn
     uint64_t& pawns = newBoard.pieces[type];
     clearPiece(pawns, move.fromPos(), type, newBoard.hash);
+    core::hashPiece(type, move.fromPos(), newBoard.kpHash);
 
     /* clear piece that will be taken if capture */
     if (move.isCapture()) {
         if (const auto victim = newBoard.getTargetAtSquare<player>(move.toSquare())) {
             clearPiece(newBoard.pieces[victim.value()], move.toPos(), victim.value(), newBoard.hash);
+
+            if (victim == WhitePawn || victim == BlackPawn) {
+                core::hashPiece(*victim, move.toPos(), newBoard.kpHash);
+            }
         }
     }
 
     switch (move.promotionType()) {
     case PromotionNone:
+        assert(false);
```

Comment:
I think we're currently mixing hard and soft error handling throughout the codebase.

e.g. `getAttackerAtSquare` fails with a bad optional access, rather than assert we've added.

Do you think it's worth giving a moment's thought towards standardizing our error handling in a lightweight way?

Reply:
Assertions are free (no runtime penalty). Ofc only for release builds - which is what we optimize for.

We shouldn't overdo it, but there's absolutely no harm in adding assertions like these! 

Runtime error handling where not optimized away doing release builds are no go (maybe what you're refering to). It's simply too expensive. 

Your example with 'std::optional' has implicit assertion so it's the same thing :)